goto_symex: rename type symbols only when needed

kroening opened the pull request.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -788,12 +788,62 @@ void goto_symex_statet::rename_address(
+  {
+    const auto &array_type=to_array_type(type);
+    return needs_renaming(array_type.subtype(), ns) ||
+           !array_type.size().is_constant();
```
This case, and also the struct/union case are only almost the same as the actual renaming function. Such subtle differences look dangerous.

kroening replied:
Fixed by means of an extra commit that first does cleanup.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -788,12 +788,62 @@ void goto_symex_statet::rename_address(
   levelt level)
 {
+  if(!needs_renaming(type, ns))
+    return;
```
That will fail to do `l1_type_entry.first->second=type;`

kroening replied:
Fixed.

kroening committed:
refactoring of goto_symex_state::rename()

kroening committed:
goto_symex: rename type symbols only when needed

Renaming is rarely needed: only if the struct contains a variable-sized
member, which is a gcc extension.  The key benefit of not renaming in all
other cases is that struct and union tag types no longer get expanded.

tautschnig reviewed (changes requested):
Well, it looked ok - but tests are failing, notably tests involving bit fields.

kroening committed:
bugfix: must use base_type_eq to compare types

kroening committed:
string refiner needs to use follow when testing types

kroening committed:
added a variant of has_subtype that takes a predicate that is given a namespace

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -689,14 +683,14 @@ void goto_symex_statet::rename(
-    const symbolt &symbol = ns.lookup(to_union_tag_type(type));
-    type = symbol.type;
+    const auto &union_type = ns.follow_tag(to_union_tag_type(type));
+    type = union_type;
```
it seems `union_type` is not used aside from that copy, so type can be copied directly from the `follow_tag` result

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -689,14 +683,14 @@ void goto_symex_statet::rename(
-    const symbolt &symbol = ns.lookup(to_struct_tag_type(type));
-    type=symbol.type;
+    const typet &struct_type = ns.follow_tag(to_struct_tag_type(type));
+    type = struct_type;
```
same thing as above

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -660,22 +660,16 @@ void goto_symex_statet::rename(
   }
-  else if(type.id()==ID_struct ||
-          type.id()==ID_union ||
-          type.id()==ID_class)
```
This is not refactoring as the commit message suggests and this change causes some tests (in particular jbmc/regression/jbmc-strings/StringSubstring) to fail.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -660,22 +660,16 @@ void goto_symex_statet::rename(
-      // be careful, or it might get cyclic
-      if(it->type().id()==ID_array)
-        rename(to_array_type(it->type()).size(), ns, level);
-      else if(it->type().id()!=ID_pointer)
```
This change doesn't look like refactoring

tautschnig commented:
@kroening Is this still relevant?

TGWDB commented:
Closing as this has been untouched for a long time and no response to author query circa Feb. 2019. Please reopen if you believe this is erroneous.
